lib/libstorj/src/storj.h

```diff
@@ -0,0 +1,905 @@
+    return value;\
+}\
+
+#define STORJ_RETURN_SET_STATE_ERROR_IF_LAST_ERROR \
+if (strcmp("", *STORJ_LAST_ERROR) != 0) { \
+    state->error_status = STORJ_LIBUPLINK_ERROR; \
+    return;\
+}\
+
+// TODO: should req->status_code be an http error status code?
+// (look into how req->status_code is used)
+#define STORJ_RETURN_SET_REQ_ERROR_IF_LAST_ERROR \
+if (strcmp("", *STORJ_LAST_ERROR) != 0) { \
+    req->error_code = STORJ_LIBUPLINK_ERROR; \
+    req->status_code = 1; \
+    return;\
+}\
+
+// TODO: do we need `extern`?
+extern char **STORJ_LAST_ERROR;
+
+typedef enum {
+    STORJ_REPORT_NOT_PREPARED = 0,
+    STORJ_REPORT_AWAITING_SEND = 1,
+    STORJ_REPORT_SENDING = 2,
+    STORJ_REPORT_SENT = 3
+} exchange_report_status_t;
+
+/** @brief Bridge configuration options
+ *
+ * Proto can be "http" or "https", and the user/pass are used for
+ * basic authentication to a Storj bridge.
+ */
+typedef struct {
+    char *addr;
+    // NB: apikey is project-specific.
+    char *apikey;
+
+} storj_bridge_options_t;
+
+/** @brief File encryption options
+ *
+ * The mnemonic is a BIP39 secret code used for generating keys for file
+ * encryption and decryption.
+ */
+typedef struct storj_encrypt_options {
+    uint8_t key[32];
+} storj_encrypt_options_t;
+
+
+
+/** @brief HTTP configuration options
+ *
+ * Settings for making HTTP requests
+ */
+typedef struct storj_http_options {
+    const char *user_agent;
+    const char *proxy_url;
+    const char *cainfo_path;
+    uint64_t low_speed_limit;
+    uint64_t low_speed_time;
+    uint64_t timeout;
+} storj_http_options_t;
+
+/** @brief A function signature for logging
+ */
+typedef void (*storj_logger_fn)(const char *message, int level, void *handle);
+
+/** @brief Logging configuration options
+ *
+ * Settings for logging
+ */
+typedef struct storj_log_options {
+    storj_logger_fn logger;
+    int level;
+} storj_log_options_t;
+
+/** @brief A function signature for logging
+ */
+typedef void (*storj_logger_format_fn)(storj_log_options_t *options,
+                                       void *handle,
+                                       const char *format, ...);
+
+/** @brief Functions for all logging levels
+ */
+typedef struct storj_log_levels {
+    storj_logger_format_fn debug;
+    storj_logger_format_fn info;
+    storj_logger_format_fn warn;
+    storj_logger_format_fn error;
+} storj_log_levels_t;
+
+/** @brief A structure for a Storj user environment.
+ *
+ * This is the highest level structure and holds many commonly used options
+ * and the event loop for queuing work.
+ */
+typedef struct storj_env {
+    storj_bridge_options_t *bridge_options;
+    storj_encrypt_options_t *encrypt_options;
+    storj_log_options_t *log_options;
+
+    uv_loop_t *loop;
+    uv_async_t async;
+    storj_log_levels_t *log;
+
+    /* New in V3 */
+    UplinkRef uplink_ref;
+    ProjectRef project_ref;
+
+    // TODO: delete?
+    /* unused in V3 */
+    storj_http_options_t *http_options;
```

Comment:
Perhaps, leave it as is here, but ignore it in the implementation.

Reply:
:+1: